File: sdks/node/src/middleware.ts
import type { OutcomeEvent } from "./events";
import type { EnqueueOptions } from "./native";

/** Context for a task as it executes on a worker. */
export interface TaskContext {
  jobId: string;
  taskName: string;
  args: unknown[];
}

/**
 * Context for a job being enqueued, passed to {@link Middleware.onEnqueue} before
 * serialization. Mutate `args`/`options` in place to validate, redact, or rewrite
 * the job; throw to abort the enqueue.
 */
export interface EnqueueContext {
  readonly taskName: string;
  /** Positional args, mutable before they are serialized. */
  args: unknown[];
  /** Enqueue options, mutable before they reach the core. */
  options: EnqueueOptions;
}

/**
 * Cross-cutting hooks around task execution and job outcomes. Register with
 * {@link Queue.use}. `onEnqueue` runs (sync) on the enqueuing side before
 * serialization; `before`/`after`/`onError` wrap execution (awaited, counted
 * toward the timeout); the outcome hooks fire after the core decides the result.
 */
export interface Middleware {
  onEnqueue?(ctx: EnqueueContext): void;
  before?(ctx: TaskContext): void | Promise<void>;
  after?(ctx: TaskContext, result: unknown): void | Promise<void>;
  onError?(ctx: TaskContext, error: unknown): void | Promise<void>;
  onCompleted?(event: OutcomeEvent): void;
  onRetry?(event: OutcomeEvent): void;
  onDeadLetter?(event: OutcomeEvent): void;
  onCancel?(event: OutcomeEvent): void;
}

File: sdks/java/src/main/java/org/byteveda/taskito/middleware/Middleware.java
package org.byteveda.taskito.middleware;

import org.byteveda.taskito.events.OutcomeEvent;

/**
 * Cross-cutting hooks around enqueue, task execution, and job outcomes. All are
 * optional. {@code onEnqueue} runs on the producer before serialization;
 * {@code before}/{@code after}/{@code onError} wrap execution; the outcome hooks
 * fire after the core decides the result. Register with {@link org.byteveda.taskito.Taskito#use}.
 */
public interface Middleware {
    default void onEnqueue(EnqueueContext context) {}

    default void before(TaskContext context) {}

    default void after(TaskContext context, Object result) {}

    default void onError(TaskContext context, Throwable error) {}

    default void onCompleted(OutcomeEvent event) {}

    default void onRetry(OutcomeEvent event) {}

    default void onDeadLetter(OutcomeEvent event) {}

    default void onCancel(OutcomeEvent event) {}
}
